F133 ServiceWorker in Workers (#36726)
mdn/content: files/en-us/web/api/serviceworkercontainer/register/index.md

{{APIRef("Service Workers API")}}{{SecureContext_Header}}{{AvailableInWorkers}}

The **`register()`** method of the {{domxref("ServiceWorkerContainer")}} interface creates or updates a {{domxref("ServiceWorkerRegistration")}} for the given scope.
If successful, the registration associates the provided script URL to a _scope_, which is subsequently used for matching documents to a specific service worker.

A single registration is created for each unique scope.
If `register()` is called for a scope that has an existing registration, the registration is updated with any changes to the scriptURL or options.
If there are no changes, then the existing registration is returned.

> Note that the service worker is installed only on the first registration of a particular scriptURL.

Reviewer: Again this is I think not true. (I believe what happens with updates is that the browser periodically gets the worker from its URL and compares its contents, if they differ it installs the update).

What is true is that `register()` is effectively a no-op if the script URL (and also options?) has not changed.

Author: Arrrg. What it is supposed to say is that calling registration doesn't automatically run the whole register/install lifecycle again.

CHanging the options and nothing else returns the same registration and updates it (^^_the registration is updated with any changes to the scriptURL or options._), without running the install process again. 

```suggestion
Note that calling `register()` with the same scope and `scriptURL` does not restart the installation process.
```

You can therefore call this method unconditionally from a controlled page: you don't need to first check whether there's an active registration or service worker.

A document can potentially be within the scope of several registrations with different service workers and options.
The browser will associate the document with the matching registration that has the most specific scope.
This ensures that only one service worker runs for each document.

> [!NOTE]

> > It is generally safer not to define registrations that have overlapping scopes.

Author: FYI verified by testing on chrome and FF. Wasn't completely clear to me from spec.

Essentially though the scope is what creates a unique registration. If you register again with the same scope the original is updated.

### Parameters

- `scriptURL`
  - : The URL of the service worker script.
    The registered service worker file needs to have a valid [JavaScript MIME type](/en-US/docs/Web/HTTP/MIME_types#textjavascript).
- `options` {{optional_inline}}

  - : An object containing registration options. Currently available options are:

    - `scope`

      - : A string representing a URL that defines a service worker's registration scope; that is, what range of URLs a service worker can control.

>         This is usually specified as an URL that is relative to the base URL of the site (e.g. `/some/path/`), so that the resolved scope is the same irrespective of what page the registration code is called from.

Reviewer 2: ```suggestion
        This is usually specified as a URL that is relative to the base URL of the site (e.g. `/some/path/`), so that the resolved scope is the same irrespective of what page the registration code is called from.
```

>         This is usually specified as an URL that is relative to the base URL of the site (e.g. `./some/path/`), so that the resolved scope is the same irrespective of what page the registration code is called from.

Author: ```suggestion
        This is usually specified as an URL that is relative to the base URL of the site (e.g. `/some/path/`), so that the resolved scope is the same irrespective of what page the registration code is called from.
```

>         By default, the `scope` value for a service worker registration is set to the directory where the service worker script is located (by resolving `./` against `scriptURL`).

Author: This was an important fix, and it flows to the rest of the examples.

Previously this whole document was a very confusing because it mentioned application-relative URLs here, and then in the examples all the URLs are page-relative BUT in most cases they are in the root so it makes no difference. This made the examples confusing.

My take on this is that all URLS should be site-relative, and I have changed the examples appropriately. If you do page-relative URLs for script and scope you're screwed if the page moves or you want to call your registration from more than one place - because you'd have to modify the URLs to be correct relative path for every location. Crazy.

        The scope must specify documents that are in the same directory or more deeply nested than the service worker (if you need a broader scope, this can be permitted via the HTTP `Service-Worker-Allowed` header).

        See the [Examples](#examples) section for more information on how it works.

    - `type`

      - : A string specifying the type of worker to create.
        Valid values are:

        - `'classic'`
          - : The loaded service worker is in a standard script.
            This is the default.
        - `'module'`
          - : The loaded service worker is in an [ES module](/en-US/docs/Web/JavaScript/Guide/Modules) and the import statement is available on worker contexts.
            For ES module compatibility info, see the [browser compatibility data table for the `ServiceWorker` interface](/en-US/docs/Web/API/ServiceWorker#browser_compatibility).

    - `updateViaCache`

>       - : A string indicating how the HTTP cache is used for service worker scripts resources during updates.

Reviewer 2: ```suggestion
      - : A string indicating how the HTTP cache is used for service worker script resources during updates.
```

### Return value

A {{jsxref("Promise")}} that resolves with a {{domxref("ServiceWorkerRegistration")}} object.

### Exceptions

- `TypeError`

  - : The `scriptURL` or `scope URL` is a failure.
    This can happen if the URL can't be resolved to a valid URL or uses a scheme that is not `http:` or `https`.

>     For `scriptURL` it may also happen because the URL cannot be validated using the [Trusted Types API](/en-US/docs/Web/API/Trusted_Types_API).

Reviewer: Would it be correct to say: this may happen if the site require the use of trusted types, but `scriptURL` was not a trusted type and could not be converted to a trusted type? https://w3c.github.io/trusted-types/dist/spec/#abstract-opdef-get-trusted-type-compliant-string.

Not sure. I feel there's a lot to unpack in here!

Author: Yes. Though I'm not sure about the detail of conversion to trusted types. I had deliberately chosen not to go to that level of detail. I'll be a little more explicit though

```suggestion
    It may also happen if `scriptURL` is not a {{domxref("TrustedScriptURL")}}, and this is a requirement of the site's [Trusted Types Policy](/en-US/docs/Web/API/Trusted_Types_API).
```

Reviewer: Yeah, I can see a lot more trusted types work in both our futures, and we can dig into this more then.

    The exception is also raised if the `scriptURL` or `scope URL` path contains the case-insensitive ASCII "%2f" (`*`) or "%5c" (`=`)

- `SecurityError` {{domxref("DOMException")}}

>   - : The `scriptURL` is not a potentially trustworthy origin, such as `localhost` or an `https` URL.

Reviewer: I don't understand how we throw one exception if the scheme is http or https (see above) when actually only https is allowed.

Author: Yeah, this is odd. Reading through the spec the checks are first on http and https which gives you a typeerror, Then later on there is a check on it being a potentially trustworthy origin.

So to answer your question, localhost considered a potentially trustworthy URL even if it is the normal http localhost.
So http is allowed in localhost.

Does that make sense?

Author: Note, I don't talk about these requirements up in `scriptURL` because they are implied by the secure context header.

}
```

Note that we have registered the scriptURL relative to the site root rather than the current page.
This allows the same registration code to be used from any page.

### Register a service worker with an explicit default scope

The code below is almost identical, except we have specified the scope explicitly using `{ scope: "/" }`.
Again, we've specified the scope as site-relative so the same registration code can be used from anywhere in the site.

```js
if ("serviceWorker" in navigator) {
  // declaring scope manually
  navigator.serviceWorker.register("./sw.js", { scope: "/" }).then(
    (registration) => {
      console.log("Service worker registration succeeded:", registration);
    },
    (error) => {
      console.error(`Service worker registration failed: ${error}`);
    },
  );
} else {
  console.error("Service workers are not supported.");
}
```

This scope happens to be the same as the default scope, so the registration applies to exactly the same pages as the example above.

> Note that if we were to run this code after the previous example, browsers should recognise that we're updating an existing registration rather than a new one.

Reviewer 2: ```suggestion
Note that if we were to run this code after the previous example, browsers should recognize that we're updating an existing registration rather than a new one.
```

### Register a service worker using page-relative URLs

> There is nothing to stop you using page-relative URLs except that this makes it harder to move your pages around, and it is easy to accidentally create unwanted registrations if you do so.

Reviewer 2: ```suggestion
There is nothing to stop you from using page-relative URLs except that this makes it harder to move your pages around, and it is easy to accidentally create unwanted registrations if you do so.
```
